### DISCUSSION TITLE
RFC: Eager <head>

### DISCUSSION THREAD

**Discussion by mohsen1** [vercel/next.js#16854] [Ideas]
## Problem
This is an idea that came to me after trying to solve a problem I faced at work where the `getInitialProps` call takes too long and it was holding the HTTP response from starting to stream. I have screenshots and more details here 

https://github.com/vercel/next.js/discussions/16730

In short, when server is busy computing the response data we can take send part of the document (the `<head>`) to let the client start downloading assets.

## Suggestion: Eager `<head>`
Taking advantage of [chunked transfer encoding](https://en.wikipedia.org/wiki/Chunked_transfer_encoding) we can render the document `<head>` before rendering entire document. In a separate `ReactDOM.renderToString` call we render the document `<head>` and start the response stream with it. Once `getInitialProps` (and other data fetching functions) of Document is completed, we render the document again and continue the response stream with the rest of document.

Document rendered string should not contain the `<head>` element. 

### `page/_head.tsx`
To simplify things we can introduce a new special page that contains all static `<head>` content. This will allow for dynamic `<head>` with props and custom `getServerSideProps` that is separate from the main Document `getServerSideProps`. This page will be optional just like `_app` and `_document`. If this page exist, Next.js will use it to render the initial response stream. 

## Challenges
This is not a trivial change. Next.js always assumed that the document is rendered and sent in one shot. 

### `<head>` with props 
Next.js allows users to make dynamic `<head>` part of the document by allowing props to be passed to the `<head>` part of `_document`. If all props for `_document` is computed in one function, the head can not be produced before computing all props which is the main rationale for this proposal. Introducing `pages/_head.tsx` can solve this problem. 

### `next-server` code complexity 
Everything in next-server from `renderToHTMLWithComponents` to `sendPayload` is assuming single chunk responses. Changing code to reflect this design would be quite difficult. 

### Content dependent headers
Some headers depend on the content. For example  we need the entire response to compute the `ETag` header. With chunked encoding we can take advantage of [HTTP Trailers](https://nodejs.org/docs/latest/api/http.html#http_response_addtrailers_headers) to send those headers after response stream is drained. 

Problem with Trailers is that they have poor tooling. Chrome dev tools and cURL for instance don't show them by default ([Chrome bug](https://bugs.chromium.org/p/chromium/issues/detail?id=422958))

## Implementation roadmap

First, to understanding chunked encoding better, take a look at the example server I put together here:

https://github.com/mohsen1/straming-http-example/blob/master/server.js

This should be an experimental feature for a while. It's a big change and can break many things. We can also do this only if `pages/_head.tsx` is available in userland. 

## Alternatives

I considered H/2 Push initially to solve this problem. I'm open to know if there are more alternatives.

### HTTP/2 Push
Alternative to HTTP streaming we can could depend on HTTP/2 Push to push out assets. I looked into H/2 Push and it feels like an overkill. H/2 Push is extremely hard to configure. The browser cache is ignored when pushing assets with HTTP/2 Push. We had to implement complex caching logic to enable HTTP/2 Push, otherwise we could waste bandwidth pushing assets that browser already has in cache. 

### 103 Early Hints and `Link` header
[103 Early Hints](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/103) status code allows for sending `Link` headers that will hint to the client to download various assets. This can help us avoid streaming chunked responses altogether but browser support is probably not good. 

Benefit over H/2 Push is that server doesn't have to do cache management. Browsers will decide to download assets based on their cache management policies.

Drawbacks:

- Early Hints are not even supported in Chrome yet. [Chrome issue](https://bugs.chromium.org/p/chromium/issues/detail?id=671310)
- It's only safe to use this over HTTP/2 otherwise very old clients that don't support HTTP/2 will break

**Comment by c0b41** [comment]
we already have similar RFC #8981

**Comment by azylinski** [comment]
`103 Early Hints` are now (since version 103) supported in Chrome: https://developer.chrome.com/blog/new-in-chrome-103/#http103

**Comment by BleedingDev** [comment]
Any update on this? It would be awesome to send early hints. :)

**Comment by theodorebear** [comment]
i love the idea of Early Hints natively in nextjs. 
faster loading, lighthouse speeds, less FOUT, whats not to like? 
thank you to the amazing NextJS team!!
